### PR TITLE
[ENH] Added `TimeMoE` forecaster to ptf-v2

### LINKED ISSUE #2137: [ENH] Add TimeMOE Model (THUML) to ptf-v2
**Is your feature request related to a problem? Please describe.**

This issue proposes adding the **TimeMOE (Mixture-of-Experts for Time Series)** model from the THUML repository to `ptf-v2` in `sktime`.

**Describe the solution you'd like**
Add the following files.
- `_timemoe_v2.py`
- `_timemoe_pkg_v2.py`

**Additional context**

- TimeMOE GitHub: https://github.com/thuml/Time-Series-Library/blob/main/models/TimeMoE.py

I would like to work on this estimator. 
CC: @PranavBhatP @phoeenniixx @fkiraly

### PR DESCRIPTION
**Pull request by lucifer4073**
#### Reference Issues/PRs

Closes #2137.

#### What does this implement/fix? Explain your changes.

Implementation adds the `TimeMoE` model in ptf-v2 inheriting `TSLibBaseModel`. Most of the implementation is similar to that in Thuml Time series library, except the following changes
- `training` argument which if set to false freezes the model backbone and uses `.generate()` method for zero-shot-inference. Defaults to True and passes data directly to backbone.
- used `logits` to get `future_flat`.

#### What should a reviewer concentrate their feedback on?

Implementation adds `transformer` dependency to ptf-v2.

#### Did you add any tests for the change?

`pkg` file consists of all the test args.

#### PR checklist

- [X] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.
- [X] Added/modified tests
- [X] Used pre-commit hooks when committing to ensure that code is compliant with hooks. Install hooks with `pre-commit install`.
  To run hooks independent of commit, execute `pre-commit run --all-files`

### CODE REVIEW

**Comment by PranavBhatP** [comment]
@lucifer4073 are you sure about interfacing `TimeMoE` as a child class of `TslibBaseModel`? If I'm not wrong it's a foundational model and that would require an entirely different D2 layer, whose design hasn't been consolidated or discussed, please suggest a prior design before making large PRs. It gets very hard to review without prior context or discussion on the design. There is parallel work going on with other foundational models like Chronos for v2 you can catch up with that discussion, and see whether your current design is feasible

**Comment by lucifer4073** [comment]
> @lucifer4073 are you sure about interfacing `TimeMoE` as a child class of `TslibBaseModel`? If I'm not wrong it's a foundational model and that would require an entirely different D2 layer, whose design hasn't been consolidated or discussed, please suggest a prior design before making large PRs. It gets very hard to review without prior context or discussion on the design. There is parallel work going on with other foundational models like Chronos for v2 you can catch up with that discussion, and see whether your current design is feasible

@PranavBhatP Thanks for the comment. I was under the impression that all models in the THUML Time-Series module inherit from `TslibBaseModel`.  
Regarding `TimeMoE`, it primarily leverages Hugging Face’s `AutoModelForCausalLM` for training and inference, and from what I can see, it aligns well with the existing TSlib design (with added dependency).
That said, I’ll review the Chronosv2 discussion more carefully to determine whether it would require a separate or entirely different D2 layer implementation. Thanks again for pointing this out.

**Comment by PranavBhatP** [comment]
> I was under the impression that all models in the THUML Time-Series module inherit from TslibBaseModel.

Yes, you are right, but the "models" here are only trainable encoder decoder models, not pre-trained models with imported weights. There are already some models like `TimeXer` which are based on the transformer architecture, so you are not completely wrong here. I think your work on TimeMoe would definitely help find common patterns in creating a design for interfacing FMs into the library along with the work on Chronos.

**Comment by PranavBhatP** [comment]
Also just another comment, try sticking to a smaller number of PRs. It would help you prioritize efforts on quality. The emphasis is on Quality > Quantity. I see almost 4 active model interface PRs from your end, but the effort is appreciated. This is just an advice from my end :)

**Comment by lucifer4073** [comment]
> Also just another comment, try sticking to a smaller number of PRs. It would help you prioritize efforts on quality. The emphasis is on Quality > Quantity. I see almost 4 active model interface PRs from your end, but the effort is appreciated. This is just an advice from my end :)

Yes, thanks for this. Actually, two of my PRs are ready for review and all test cases are passing. One of them has already been reviewed and can be merged. So, in the meantime, I thought I could open a few more PRs for model interfacing.
